File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-0.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        left = 0
        right =len(nums) - 1

        while left < right:
            mid = (left + right) // 2
            if nums[mid] > nums[right]:
                left = mid + 1
            else:
                right = mid
            
        pivot = left

        def binarySearch(left, right):
            while left <= right:
                mid = (left + right) // 2
                if nums[mid] == target:
                    return mid
                elif nums[mid] < target:
                    left = mid + 1
                else:
                    right = mid - 1
            return -1
        
        result = binarySearch(0, pivot - 1)
        if result != -1:
            return result
        
        return binarySearch(pivot, len(nums) - 1)

File: Data Structures & Algorithms/binary-search/submission-0.java
class Solution {
    public int search(int[] nums, int target) {
        return binarySearch(0, nums.length - 1, nums, target);
    }

    private int binarySearch(int left, int right, int[] nums, int target) {
        if (left > right) {
            return -1;
        }

        int medium = left + (right - left) / 2;

        if (nums[medium] == target) {
            return medium;
        }

        return (nums[medium] < target) ? 
            binarySearch(medium + 1, right, nums, target) : binarySearch(left, medium - 1, nums, target);
    }
}
